src/persist/src/indexed/snapshot.rs

```diff
@@ -332,3 +258,87 @@ impl Iterator for ArrangementSnapshotIter {
         })
     }
 }
+
+/// A type that [BatchesIter] can iterate over.
+trait BatchesIterBatch {
+    fn chunks(&self) -> &[ColumnarRecords];
+}
+
+impl BatchesIterBatch for BlobUnsealedBatch {
+    fn chunks(&self) -> &[ColumnarRecords] {
+        &self.updates
+    }
+}
+
+impl BatchesIterBatch for BlobTraceBatch {
+    fn chunks(&self) -> &[ColumnarRecords] {
+        &self.updates
+    }
+}
+
+/// An internal helper for iterating over the result of a set of Futures
+/// (representing fetches from storage), each of which resolves to something
+/// that has a slice of [ColumnarRecords].
+#[derive(Debug)]
+struct BatchesIter<B: BatchesIterBatch> {
+    chunk_idx: usize,
+    chunks_idx: usize,
+    current: Option<Arc<B>>,
+    batches: VecDeque<PFuture<Arc<B>>>,
+}
+
+impl<B: BatchesIterBatch> Default for BatchesIter<B> {
+    fn default() -> Self {
+        Self {
+            chunk_idx: Default::default(),
+            chunks_idx: Default::default(),
+            current: Default::default(),
+            batches: Default::default(),
+        }
+    }
+}
+
+impl<B: BatchesIterBatch> Iterator for BatchesIter<B> {
+    type Item = Result<((Vec<u8>, Vec<u8>), u64, i64), Error>;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        loop {
+            let current = match self.current.as_ref() {
+                Some(x) => x,
+                None => {
+                    let new_current = match self.batches.pop_front() {
+                        Some(x) => x,
+                        None => return None,
+                    };
+                    let new_current = match new_current.recv() {
+                        Ok(x) => x,
+                        Err(err) => return Some(Err(err)),
```

Comment:
would it be better to just return None after we return the error? In other words after sending an error the iterator refuses to send any other records? Not sure if that would be better for finding issues. No need to do that before merging this change just a random thought.

Reply:
that's an interesting idea, but my initial reaction is that IMO is that feels like a less natural/more confusing contract for the iterator. if we wanted logic like that somewhere, I'd say it should either live in the caller or in some sort of iterator adaptor (analogous to the stdlib "Fuse" iterator adaptor for non-fused to fused https://doc.rust-lang.org/std/iter/struct.Fuse.html)